src/snowflake/snowpark/mock/plan.py

```diff
@@ -167,6 +189,87 @@ def __init__(
         self.options = options
 
 
+def handle_order_by_clause(
+    order_by: List[SortOrder],
+    result_df: TableEmulator,
+    analyzer: "MockAnalyzer",
+    expr_to_alias: Optional[Dict[str, str]],
+) -> TableEmulator:
+    """Given an input dataframe `result_df` and a list of SortOrder expressions `order_by`, return the sorted dataframe."""
+    sort_columns_array = []
+    sort_orders_array = []
+    null_first_last_array = []
+    for exp in order_by:
+        sort_columns_array.append(analyzer.analyze(exp.child, expr_to_alias))
+        sort_orders_array.append(isinstance(exp.direction, Ascending))
+        null_first_last_array.append(
+            isinstance(exp.null_ordering, NullsFirst) or exp.null_ordering == NullsFirst
+        )
+    for column, ascending, null_first in reversed(
+        list(zip(sort_columns_array, sort_orders_array, null_first_last_array))
+    ):
+        comparator = partial(custom_comparator, ascending, null_first)
+        result_df = result_df.sort_values(by=column, key=comparator)
+    return result_df
+
+
+def handle_range_frame_indexing(
+    order_spec: List[SortOrder],
+    res_index: "pd.Index",
+    res: "pd.api.typing.DataFrameGroupBy",
+    analyzer: "MockAnalyzer",
+    expr_to_alias: Optional[Dict[str, str]],
+    unbounded_preceding: bool,
+    unbounded_following: bool,
+) -> "pd.api.typing.RollingGroupby":
+    """Return a list of range between window frames based on the dataframe paritions `res` and the ORDER BY clause `order_spec`."""
+    if order_spec:
+        windows = []
+        for current_row, win in zip(res_index, res.rolling(EntireWindowIndexer())):
+            if order_spec:
+                _win = handle_order_by_clause(order_spec, win, analyzer, expr_to_alias)
+                row_idx = list(_win.index).index(current_row)
+                start_idx = 0 if unbounded_preceding else row_idx
+                end_idx = len(_win) - 1 if unbounded_following else row_idx
+                while start_idx > 0:
```

Comment:
should we add an upper bound for start_idx?

Reply:
`start_idx` is initiated to be either 0 or `row_idx` (guaranteed to exist), then it only decrements, so I don't think we have to?